doc/design/tensor_design.md

```diff
@@ -0,0 +1,175 @@
+# Tensor Design
+
+
+## Allocation
+
+`Allocation` is a [RAII](http://en.cppreference.com/w/cpp/language/raii) class template, which is used to handle a piece of memory. 
+
+```cpp
+// Template parametr 'Device' can be 'CpuDevice' or 'GpuDevice'
+template <typename Device>
+class Allocation {
+public:
+    Allocation();
+    Allocation(size_t size, Device device);
+
+    // Creates a non-owned allocation
+    Allocation(void* ptr, size_t size, Device device);
+
+    ~Allocation();
+    //No copying!
```

Comment:
Maybe in the `framework` directory, we could write `DISABLE_COPY` macro, because there are many classes are be `DISABLE_COPY`.

Otherwise, if we are happy to use `boost`, there should be `boost::non_copyable` base class.